src/types.ts

```diff
@@ -401,25 +401,29 @@ type ExtractLeaf<P> = P extends `${string}->>${infer Col}` ? Col : never;
  *
  * @typeParam P - path ending with `->>` with optional `->` in between
  * @typeParam O - object possibly containing the path
- * @typeParam L - leaf key in recursive calls
+ * @typeParam N - `null` that needs to be caried recursively
+ * if any of the parents could be `null`.
+ * It should be left empty in initial call.
  */
 type ExtractObj<
   P extends string,
   O extends Record<string, any>,
-  L extends keyof O = never,
-> = keyof O extends string
-  ? P extends `${infer P}->>${infer LL}`
-    ? P extends `${infer Col}->${infer PP}`
-      ? ExtractObj<PP, O[Col], LL>
-      : keyof O[P] extends string
-        ? string
-        : string | null
-    : P extends `${infer Col}->${infer PP}`
-      ? PP extends `${string}->${string}`
-        ? ExtractObj<PP, O[Col], L>
-        : string
-      : string
-  : string | null;
```

Comment:
This looked like it had excess complexity, especially because of the duplicated condition, so I found a way to simplify it
```ts
P extends `${infer Col}->${infer PP}`
```